ddclient.in

```diff
@@ -2273,65 +2273,79 @@ sub get_ip {
     return $ip;
 }
 
+
+######################################################################
+## Regex to find IPv4 address. Accepts embedded leading zeros.
+######################################################################
+my $regex_ipv4 = qr/(?:(?<octet>25[0-5]|2[0-4]\d|[01]?\d\d?)\.){3}(?&octet)/;
+
+######################################################################
+## Regex to find IPv6 in standard, compressed format. Accepts
+## embedded leading zeros in the address. Adapted from regex in the
+## Regular Expressions Cookbook by J.Goyvaerets and S.Levithan
+######################################################################
+my $regex_ipv6_only = qr/(?<![:.0-9A-Z])        ## Negative lookbehind for word boundry
+    (?:[0-9A-F]{1,4}:){7}[0-9A-F]{1,4}          ## Is in standard format
+      |(?=(?:[0-9A-F]{0,4}:){0,7}[0-9A-F]{0,4}  ## OR compressed with at most 7 colons
+       (?![:.0-9A-Z]))                          ##   and negative lookahead for boundry
+       (?:(?:[0-9A-F]{1,4}:){1,7}|:)(?:(?::[0-9A-F]{1,4}){1,7}|:)  ## and only 1 double-colon
```

Comment:
This case in particular is really difficult to read. I know it's correct, but I think we can write something that's easier to understand. What do you think about this:

```perl
my $regex_ipv6_only = qr/
  # Define some named groups so we can use Perl's recursive subpattern feature for shorthand:
  (?<g>[0-9A-F]{1,4}){0}  # "g" matches a group of 1 to 4 hex chars
  (?<g_>(?&g):){0}        # "g_" matches a group of 1 to 4 hex chars followed by a colon
  (?<_g>:(?&g)){0}        # "_g" matches a colon followed by a group of 1 to 4 hex chars
  (?<g0>(?&g)?){0}        # "g0" is an optional "g" (matches a group of 0 to 4 hex chars)
  (?<g0_>(?&g0):){0}      # "g0_" is an optional "g" followed by a colon
  (?<x>[:.0-9A-Z]){0}     # "x" matches chars that should never come before or after the address

  # Now for the actual regular expression:

  # Not preceded by a character that is not allowed to come before the address
  (?<!(?&x))
  (?:
    # Exactly 8 groups of 1-4 hex chars
    (?&g_){7}(?&g)
    # OR compressed form with the double colon at the end
    | (?&g_){1,7}:
    # OR compressed form with the double colon at the beginning
    | :(?&_g){1,7}
    # OR compressed form with the double colon at both ends :)
    | ::
    # OR compressed form with the double colon in the middle
    | (?=                   # Only consider this case if the string...
        (?&g0_){2,7}(?&g0)  # ...has 3 to 8 possibly empty groups (at least 3 because there
                            # will be at least one group before the ::, at least one group
                            # after the ::, and the :: itself is an empty group)...
        (?!(?&x))           # ...then ends.
      )
      (?&g_){0,6}(?&g)  # 1 to 7 non-empty groups before the double colon
      ::
      (?&g_){0,6}(?&g)  # 1 to 7 non-empty groups after the double colon
  )
  # Not followed by a character that is not allowed to come after the address
  (?!(?&x))
/xi;
```

Reply:
Thats an interesting idea.  Let me run it through my tests.

Reply:
So that passes all my unit tests.  But interestingly Regex101 thinks there is an error in the pattern [(see here)](https://regex101.com/r/uG5k0D/1) but debuggex is quite [happy](https://www.debuggex.com/r/TFvjD9Kf8iT64BOK) with it.  Looks like you broke Regex101 !!

Reply:
One possible downside to my suggested regex is this bit from [perl5200delta](https://perldoc.perl.org/perl5200delta.html#Regular-Expressions) (changelog for Perl v5.20):

> Under certain conditions, Perl would throw an error if in a lookbehind assertion in a regexp, the assertion referred to a named subpattern, complaining the lookbehind was variable when it wasn't. This has been fixed. [perl #120600], [perl #120618]. The current fix may be improved on in the future.

This sounds like the Regex101 error. I don't think we need to do anything unless some of the CI runs fail. If they do, we'll have to replace `(?<!(?&x))` with `(?<![:.0-9A-Z])`.

Reply:
So you got me worried there because MacOS has 5.18.  So I just ran this through my tests on my Mac and it passed.  But we probably need to test on older versions of MacOS / OS X.  I'm looking at my other regex (the one that handles mixed IPv6/IPv4) to adopt your idea into that.  I've tested pulling out the named group definitions into their own `qr/.../` and then referencing that in the main regex's and it works 👍

Reply:
@rhansen You totally inspired me.  Check out the last commit.